src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java

```diff
@@ -363,22 +362,46 @@ else if (abbrStarted) {
     /**
      * Get Abbreviation if it is illegal.
      * @param str name
-     * @param beginIndex begin index
-     * @param endIndex end index
-     * @return true is abbreviation is bigger that required and not in ignore list
+     * @param beginIndex begin index, inclusive
+     * @param endIndex end index, inclusive
+     * @return the abbreviation if bigger than required and not in ignore list
      */
     private String getAbbreviationIfIllegal(String str, int beginIndex, int endIndex) {
         String result = null;
         final int abbrLength = endIndex - beginIndex;
         if (abbrLength > allowedAbbreviationLength) {
-            final String abbr = str.substring(beginIndex, endIndex);
+            final String abbr = getAbbreviation(str, beginIndex, endIndex);
             if (!allowedAbbreviations.contains(abbr)) {
                 result = abbr;
             }
         }
         return result;
     }
 
+    /**
+     * Gets the abbreviation, only including the character at {@code endIndex} if
+     * it is the last character in the string, as it is usually the first capital in
+     * the next word.
+     * <p>
+     * For example, {@code getAbbreviation("getXMLParser", 3, 6)} returns "XML"
+     * (not "XMLP"), and so does {@code getAbbreviation("parseXML", 5, 7)}.
```

Comment:
> getXMLParser", 3, 6) returns "XML" (not "XMLP")
> and so does getAbbreviation("parseXML"

`parseXML` should only return `XML` but `getXMLParser` should still return `XMLP` as this was the old behavior, correct?
We are counting consecutive uppercase letters. We can't possibly know if the last letter is part of a next word or not without scanning some type of dictionary.

Reply:
I agree that we can't know for sure if the last letter is part of a next word or not, but I believe that is old/existing behavior (see the comment at [line 347](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java#L347): `// -1 as a first capital is usually beginning of next word`)

The existing tests demonstrate this. For example:
- [testTypeAndVariablesAndMethodNames](https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheckTest.java#L92) checks that `marazmaticVARIABLEName` is flagged as a warning.
- [testTypeAndVariablesAndMethodNamesWithNoIgnores](https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheckTest.java#L117) checks the same file but adds `VARIABLE` as an allowed abbreviation, and then `marazmaticVARIABLEName` is no longer flagged - even though the "abbreviation" is really `VARIABLEN`.

When I first looked at this in the code I thought, that's kind of strange... but, I guess it's a reasonable heuristic and probably works right most of the time? Like, in `getXMLParser` the intended abbreviation really is `XML` and not `XMLP`.

Reply:
@esilkensen Thanks. I see now. To me it seems slightly odd, but since it is the old behavior it must stay. It slightly does make sense, but can't always be trusted.

This section can be ignored for now.